datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -530,31 +527,30 @@ fn coerce_window_frame(
         }
         WindowFrameUnits::Rows | WindowFrameUnits::Groups => &DataType::UInt64,
     };
-    window_frame.start_bound =
-        coerce_frame_bound(target_type, &window_frame.start_bound)?;
-    window_frame.end_bound = coerce_frame_bound(target_type, &window_frame.end_bound)?;
+    window_frame.start_bound = coerce_frame_bound(target_type, window_frame.start_bound)?;
+    window_frame.end_bound = coerce_frame_bound(target_type, window_frame.end_bound)?;
     Ok(window_frame)
 }
 
 // Support the `IsTrue` `IsNotTrue` `IsFalse` `IsNotFalse` type coercion.
 // The above op will be rewrite to the binary op when creating the physical op.
-fn get_casted_expr_for_bool_op(expr: &Expr, schema: &DFSchemaRef) -> Result<Expr> {
+fn get_casted_expr_for_bool_op(expr: Expr, schema: &DFSchemaRef) -> Result<Expr> {
     let left_type = expr.get_type(schema)?;
     get_input_types(&left_type, &Operator::IsDistinctFrom, &DataType::Boolean)?;
-    cast_expr(expr, &DataType::Boolean, schema)
+    expr.cast_to(&DataType::Boolean, schema)
 }
 
 /// Returns `expressions` coerced to types compatible with
 /// `signature`, if possible.
 ///
 /// See the module level documentation for more detail on coercion.
 fn coerce_arguments_for_signature(
-    expressions: &[Expr],
+    expressions: Vec<Expr>,
```

Comment:
this will stop all the arguments to function calls being copied, which seems good

Reply:
I always thought having less possible coercible type is preferred 🤔

Reply:
I am not sure what you mean

Perhaps you are referring to preferring `&[Expr]` over `&Vec<Expr>`? If so the difference here is that the Vec was owned (and thus we don't have to copy the contents via `to_vec()`)